salt/utils/vmware.py

```diff
@@ -158,104 +166,217 @@ def esxcli(host, user, pwd, cmd, protocol=None, port=None, esxi_host=None):
     return ret
 
 
-def get_service_instance(host, username, password, protocol=None, port=None):
-    '''
-    Authenticate with a vCenter server or ESX/ESXi host and return the service instance object.
-
-    host
-        The location of the vCenter server or ESX/ESXi host.
-
-    username
-        The username used to login to the vCenter server or ESX/ESXi host.
-
-    password
-        The password used to login to the vCenter server or ESX/ESXi host.
-
-    protocol
-        Optionally set to alternate protocol if the vCenter server or ESX/ESXi host is not
-        using the default protocol. Default protocol is ``https``.
-
-    port
-        Optionally set to alternate port if the vCenter server or ESX/ESXi host is not
-        using the default port. Default port is ``443``.
-    '''
-    if protocol is None:
-        protocol = 'https'
-    if port is None:
-        port = 443
-
-    service_instance = GetSi()
-    if service_instance:
-        if service_instance._GetStub().host == ':'.join([host, str(port)]):
-            if salt.utils.is_proxy():
-                service_instance._GetStub().GetConnection()
-            return service_instance
-        Disconnect(service_instance)
-
+def _get_service_instance(host, username, password, protocol,
+                          port, mechanism, principal, domain):
+    '''
+    Internal method to authenticate with a vCenter server or ESX/ESXi host
+    and return the service instance object.
+    '''
+    log.debug('Retrieving new service instance')
+    token = None
+    if mechanism == 'userpass':
+        if username is None:
+            raise excs.CommandExecutionError('Login mechanism userpass was '
+                                             'specified but the mandatory '
+                                             'parameter \'username\' is '
+                                             'missing')
+        if password is None:
+            raise excs.CommandExecutionError('Login mechanism userpass was '
+                                             'specified but the mandatory '
+                                             'parameter \'password\' is '
+                                             'missing')
+
+        if domain and '@' not in username and '\\' not in username:
```

Comment:
Why is this needed? Having multiple ways of specifying username in the cloud provider configuration would cause confusion. Also, specifying domain is not needed for ``userpass`` mechanism.